Цель работы:

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.


# Загрузка набора данных
## Описание предметной области
Вариант №5
<br>Набор данных visits.csv <br> Атрибуты:
<br>User_Id — уникальный идентификатор пользователя
<br>Region — страна пользователя
<br>Device — устройство пользователя
<br>Channel — идентификатор рекламного источника, из которого пришел пользователь
<br>Session_Start — дата и время начала сессии
<br>Session_End — дата и время окончания сессии

---
## 1. Чтение файла (набора данных)
Импорт библиотек, чтение файла с помощью pandas

In [74]:
import pandas as pd
df = pd.read_csv("../data/visits.csv", sep =';')

---
## 2. Обзор данных
 2.1 Вывод первых 20 строк с помощью метода head

In [75]:
df.head(20)

User_Id         Region   Device       Channel     Session_Start  \
0   9,81449E+11  United States   iPhone       organic   01.05.2019 2:36   
1   2,78966E+11  United States   iPhone       organic   01.05.2019 4:46   
2   5,90706E+11  United States      Mac       organic  01.05.2019 14:09   
3   3,26434E+11  United States  Android        TipTop   01.05.2019 0:29   
4   3,49774E+11  United States      Mac       organic   01.05.2019 3:33   
5   43958116050  United States  Android       organic   01.05.2019 9:03   
6   1,85365E+11  United States   iPhone       organic   01.05.2019 9:37   
7   1,00971E+11  United States      Mac        TipTop   01.05.2019 4:39   
8   3,70456E+11  United States   iPhone       organic  01.05.2019 14:44   
9   1,41838E+11  United States      Mac      FaceBoom   01.05.2019 6:20   
10  9,24309E+11  United States   iPhone       organic  01.05.2019 20:03   
11  7,74938E+11  United States   iPhone  MediaTornado  01.05.2019 14:01   
12  2,44879E+11  United States   iPhone  MediaTornado   01.05.2019 1:59   
13  1,57877E+11  United States      Mac       organic  01.05.2019 10:25   
14  5,24629E+11  United States      Mac       organic  01.05.2019 17:06   
15  1,36227E+11  United States       PC       organic  01.05.2019 19:34   
16  6,84979E+11  United States   iPhone        TipTop   01.05.2019 1:51   
17  9,05776E+11  United States   iPhone       organic  01.05.2019 14:26   
18  15174034813  United States   iPhone      FaceBoom  01.05.2019 15:34   
19  1,88246E+11  United States  Android      FaceBoom  01.05.2019 23:09   

         SESSION_End  
0    01.05.2019 2:45  
1    01.05.2019 4:47  
2   01.05.2019 15:32  
3    01.05.2019 0:54  
4    01.05.2019 3:57  
5   01.05.2019 10:08  
6   01.05.2019 10:00  
7    01.05.2019 4:57  
8   01.05.2019 15:41  
9    01.05.2019 6:54  
10  01.05.2019 20:19  
11  01.05.2019 14:29  
12   01.05.2019 2:21  
13  01.05.2019 10:35  
14  01.05.2019 18:05  
15  01.05.2019 19:39  
16   01.05.2019 2:12  
17  01.05.2019 14:35  
18  01.05.2019 15:56  
19  01.05.2019 23:21

---
 2.2 Оценка данных с помощью метода info

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 957 entries, 0 to 956
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   User_Id        957 non-null    object
 1   Region         956 non-null    object
 2   Device         955 non-null    object
 3   Channel        957 non-null    object
 4   Session_Start  955 non-null    object
 5   SESSION_End    955 non-null    object
dtypes: object(6)
memory usage: 45.0+ KB


---
2.3 Оценка данных с помощью метода describe.

In [77]:
df.describe()

User_Id         Region  Device  Channel     Session_Start  \
count           957            956     955      957               955   
unique          865              2       6        4               820   
top     1,79085E+11  United States  iPhone  organic  02.05.2019 20:16   
freq              4            955     421      612                 4   

             SESSION_End  
count                955  
unique               836  
top     02.05.2019 22:48  
freq                   3

---
Была произведена оценка данных методами describe и info. В датасете 957 строк и 6 столбцов, тип всех столбцов строковый, пропуски данных в столбцах Region(1), Device(2), Session_Start(2), Session_End(2).
***


2.4 Оценка названий столбцов

In [78]:
df.columns

Index(['User_Id', 'Region', 'Device', 'Channel', 'Session_Start',
       'SESSION_End'],
      dtype='object')

2.5 Анализ и коррекция названий столбцов
На экран были выведены названия столбцов, ошибка есть только в название последнего столбца SESSION_End, слово целиком написано в верхнем регистре, для удобства приведем все к единому виду.

In [79]:
df.columns = df.columns.str.lower()
df.columns

Index(['user_id', 'region', 'device', 'channel', 'session_start',
       'session_end'],
      dtype='object')

---
### 3. Проверка пропусков

In [80]:
df.isna().sum()

user_id          0
region           1
device           2
channel          0
session_start    2
session_end      2
dtype: int64

In [81]:
df['region'] = df['region'].fillna("United States")
df['device'] = df['device'].fillna("iPhone")
df['session_start'] = df['session_start'].fillna("01.05.2019 2:36")
df['session_end'] = df['session_end'].fillna("01.05.2019 2:45")

In [82]:
df.isna().sum()

user_id          0
region           0
device           0
channel          0
session_start    0
session_end      0
dtype: int64

---
Был выбран способ заполнения пропущенных значений значением по умолчанию, так как удаление пустых строк может привести к ошибкам в дальнейшей работе.
***

### 4. Проверка дубликатов
4.1 Проверка явных дубликатов

In [83]:
df[df.duplicated()]

user_id         region device  channel    session_start  \
480  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   
481  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   

         session_end  
480  01.05.2019 7:49  
481  01.05.2019 7:49

---
Команда `df.duplicated()` проверяет, есть ли дубликаты строк в таблице. Возвращает колонку из True и False
***

In [84]:
df = df.drop_duplicates().reset_index(drop = True)
print(df.duplicated().sum())

0


---
Команда  `df = df.drop_duplicates().reset_index(drop = True)` удаляет дубликаты и обновляет индексы строк
***

4.2 Проверка неявных дубликатов

In [85]:
unique_c = ['region', 'device', 'channel']
for i in unique_c:
  print(i, df[i].unique())

region ['United States' 'USA']
device ['iPhone' 'Mac' 'Android' 'PC' 'MAC' 'IPHONE']
channel ['organic' 'TipTop' 'FaceBoom' 'MediaTornado']


Обнаружены неявные дубликаты: <br>
 `region` - United States : USA; <br>
` device` - 'IPHONE' : 'Iphone'; 'MAC' : 'Mac'.
Дубликаты требуют устранения для дальнейшей обработки данных

In [86]:
df['region'] = df['region'].replace('USA', 'United States')
df['device'] = df['device'].replace('MAC', 'Mac')
df['device'] = df['device'].replace('IPHONE', 'iPhone')
for i in unique_c:
  print(i, df[i].unique())

region ['United States']
device ['iPhone' 'Mac' 'Android' 'PC']
channel ['organic' 'TipTop' 'FaceBoom' 'MediaTornado']


С помощью замены данные приведены к единному виду
***

### 5. Проверка типов данных

In [87]:
df.dtypes

user_id          object
region           object
device           object
channel          object
session_start    object
session_end      object
dtype: object

Проверка показала, что pandas присвоил всем данных тип object, что неправильно, это нужно исправить: <br>
`region`, `device`, `channel`  - переводим в category, так как число записей ограничено; <br>
`session_start и session_end` - переводим в datetime (тип данных для времени).

In [88]:
df['region'] = df['region'].astype('category')
df['device'] = df['device'].astype('category')
df['channel'] = df['channel'].astype('category')

df['session_start'] = pd.to_datetime(df['session_start'], format='%d.%m.%Y %H:%M')
df['session_end'] = pd.to_datetime(df['session_end'], format='%d.%m.%Y %H:%M')

print(df.dtypes)

user_id                  object
region                 category
device                 category
channel                category
session_start    datetime64[ns]
session_end      datetime64[ns]
dtype: object


### 6. Группировка данных
#### Задание 1
Группировка - device и количество рекламных источников каждого типа (channel)

In [89]:
print(df.groupby(['device', 'channel'])['user_id'].count())

device   channel     
Android  FaceBoom         33
         MediaTornado     11
         TipTop           26
         organic         115
Mac      FaceBoom         36
         MediaTornado     10
         TipTop           24
         organic         172
PC       FaceBoom         15
         MediaTornado      2
         TipTop           14
         organic          72
iPhone   FaceBoom         65
         MediaTornado     30
         TipTop           79
         organic         251
Name: user_id, dtype: int64


C:\Users\miros\AppData\Local\Temp\ipykernel_12740\1587863489.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['device', 'channel'])['user_id'].count())



Результаты группировки данных видно, что в изначальном датасете основной источник траффика органический, Iphone доминирует среди устройств, рекламные каналы показывают разную эффективность, аудитория Android меньше Iphone, мобильные устройства используются чаще портативных
***

#### Задание 2

Группировка - region и количество устройств (device ). Создать датафрейм. Переименовать столбец с количеством в “сountˮ. Отсортировать по убыванию столбца “countˮ.

In [90]:
counts = df.groupby(['region', 'device'], observed = True).size()
df_group = counts.reset_index()
df_group.columns = ['region', 'device', 'count']
df_group = df_group.sort_values(by='count', ascending=False)
print(df_group)

          region   device  count
3  United States   iPhone    425
1  United States      Mac    242
0  United States  Android    185
2  United States       PC    103



Вся аудитория из США, самое большое количество пользователей за продуктами Apple (Iphone - 425, Mac - 242)
***

#### Задание 3

Сводная таблица (pivot_table) – уникальное количество пользователей
для каждого устройства (device). Отсортировать по убыванию количества

In [91]:
pivot_users = df.pivot_table(
    index='device',
    values='user_id',
    aggfunc='nunique',
    observed=True
).sort_values(by='user_id', ascending=False)
print(pivot_users)

         user_id
device          
iPhone       387
Mac          223
Android      169
PC            90




Продукция Apple (Iphone и Mac) - доминирует, продукт нацелен именно на эту систему, нужно улучшать качество продукта для Android и PC и поднимать популярность для пользователей этих устройств.
***

#### Задание 4

Сводная таблица (pivot_table) - уникальное количество пользователей для каждого устройства (device) - столбцы и региона- строки. Отсортировать по возрастанию столбца region.

In [92]:
pivot_users = df.pivot_table(
    index=['region','channel'],
    columns='device',
    values='user_id',
    aggfunc='nunique',
    observed=True
).sort_index(ascending=True)
print(pivot_users)

device                      Android  Mac  PC  iPhone
region        channel                               
United States FaceBoom           32   32  13      63
              MediaTornado       11    9   2      25
              TipTop             19   22  10      69
              organic           108  160  65     232


Из сводной таблицы хорошо видно, что вся аудитория это пользователи из США, явное лидерство у продукции Apple (Iphone и Mac), а платформы Android и PC сильно уступают, что указывает на возможный рост в этой сфере
***

### Вывод

В ходе работы был проведён анализ набора данных visits.csv, содержащего информацию о 957 пользовательских сессиях на цифровой платформе.
**На этапе предобработки были выполнены следующие шаги:**
* Устранены пропуски в столбцах region, device, session_start, session_end
* Преобразован тип данных для столбцов region, device, channel, session_start, session_end

**Выводы из анализа данных с помощью группировок и сводных таблиц:**
* Все пользователи из United States
* Iphone является лидером среди устройтсв, за ним идет Mac, затем Android и PC
* Наибольшее количество уникальных пользователей дублирует картину, что свидетельствует о том, что основная аудитория - пользователи Apple
* Органический траффик - основной источник для всех платформ Доп задания(3 7 12 18)
***

### Дополнительные задания
#### Задание №3
Добавить столбец - “Длительность сессии” (расчетный). Выполнить
группировку: средняя и медианная длительность сессии по стране
пользователя.

In [93]:
df['session_duration_min'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

result = df.groupby('region')['session_duration_min'].agg(
    Средняя_длительность='mean',
    Медианная_длительность='median'
).round(2)
print(result)

               Средняя_длительность  Медианная_длительность
region                                                     
United States                 29.17                    21.0


C:\Users\miros\AppData\Local\Temp\ipykernel_12740\68962011.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  result = df.groupby('region')['session_duration_min'].agg(


---
**Пояснение:** Группируем данные по региону и выводим их среднюю и медианную длительность в минутах. <br>
**Вывод:** Так как страна всего одна (United States), то узнаем результаты только для неё, средняя длительность = 29.17 минут; медианная длительность = 21 минуте.

#### Задание №7
Добавить столбец - “Длительность сессииˮ (расчетный). Создать столбец “Категория длительностиˮ (с помощью категоризации). Выделить минимум категории (низкая, высокая, средняя), фильтрацию для уровня длительность выбрать самостоятельно, аргументировать выбор. Создать сводную таблицу: средняя и медианная длительность сессии по устройству и категории длительности

In [94]:
df['duration_minutes'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

Вычисляем длительность сессии в минутах и делаем категории.
* Низкая - меньше 2-х минут, пользователь быстро ушел;
* Средняя - 2-10 минут, типичное время просмотра товаров, фильтрации, добавление в коризну;
* Высокая - >10 минут, глубокое взаимодействие: сравнение товаров, анализ описаний.

In [95]:
def categorize_session(duration):
    if duration <= 2:
        return 'Низкая'
    elif duration <= 10:
        return 'Средняя'
    else:
        return 'Высокая'

df['duration_category'] = df['duration_minutes'].apply(categorize_session)

pivot_table = df.pivot_table(
    index=['device', 'duration_category'],
    values='duration_minutes',
    aggfunc={'duration_minutes': ['mean', 'median']},
    observed = True
).round(2)

pivot_table.columns = ['mean_duration', 'median_duration']
pivot_table = pivot_table.sort_index()
print(pivot_table)

                           mean_duration  median_duration
device  duration_category                                
Android Высокая                    38.90             32.0
        Низкая                      1.20              1.0
        Средняя                     6.38              6.0
Mac     Высокая                    36.56             29.0
        Низкая                      1.13              1.0
        Средняя                     6.96              8.0
PC      Высокая                    40.90             33.0
        Низкая                      1.44              2.0
        Средняя                     6.43              7.0
iPhone  Высокая                    38.88             31.0
        Низкая                      1.29              1.0
        Средняя                     6.05              6.0


---
**Пояснение:** <br>
Берём данные о сессиях, вычисляем длительность каждой в минутах и относим её к одной из трёх поведенческих категорий — «Низкая» (≤2 мин), «Средняя» (2–10 мин) или «Высокая» (>10 мин). Затем строим сводную таблица, показывающая среднюю и медианную длительность сессий по комбинациям устройств и категорий, что позволяет быстро оценить вовлечённость пользователей на разных платформах. <br>
**Вывод:** <br>
У Iphone наибольшая средняя продолжительность сессий, при этом медиана значительно ниже среднего, что указывает на наличие длинных сессий у пользователей Iphone. У остальных устройств близкие значения и среднего и медианы; Разница среднего и медианы небольшая => распределение равномерное (маленькое влияние выбросов).
***

#### Задание №12
Добавить столбец - “Длительность сессии” (расчетный). Отфильтровать набор
данных. Выбрать только те записи, средняя длительность на которых выше
определенного числа (число выбрать самостоятельно). Для отфильтрованного
набора данных выполнить группировку: тип устройства и средняя+медианная
длительность.

In [96]:
df['duration_minutes'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

df_filtered = df[df['duration_minutes'] > 15]


result = df_filtered.groupby('device')['duration_minutes'].agg(
    Средняя_длительность='mean',
    Медианная_длительность='median'
).round(2)

result

C:\Users\miros\AppData\Local\Temp\ipykernel_12740\3787399787.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  result = df_filtered.groupby('device')['duration_minutes'].agg(


Средняя_длительность  Медианная_длительность
device                                               
Android                 44.32                    37.0
Mac                     41.20                    34.0
PC                      44.94                    36.0
iPhone                  43.84                    36.0

---
**Пояснение**: для каждой строки рассчитывается длительность сессии в минутах, выбирается порог длительности (15 минут), отбираем только вовлеченных пользователей, затем данные группируются по столбцу device и рассчитывается средняя длительность сессии для устройства и медианная длительность. <br>
**Вывод**: Самая большая средняя длительность сессии за PC, пользователи больше всего вовлечены именно на этом устройстве. Медиана заметно ниже среднего, это показывает, что длинные сессии присуствуют и они тянут среднее вверх, но большинство сессий в среднем диапазоне.
***

#### Задание 18
Добавить столбец - “Длительность сессии” (расчетный). Отфильтровать набор
данных. Выбрать только те записи, средняя длительность на которых лежит в
определенном диапазоне (нижнюю и верхнюю границы выбрать
самостоятельно) + топ 2 устройство по средней длительности сессии. Для
отфильтрованного набора данных создать сводную таблицу: канал, устройство
и медианная длительность сессии.


In [97]:
df_filtered = df[(df['session_duration_min'] >= 15) & (df['session_duration_min'] <= 60)]

Отфильтровали данные по длительности. Нижняя граница = 15, берем только вовлеченных пользователей, верхняя граница = 60, скорее всего при таком большом времени пользователь оставил вкладку открытой и не взаимодействует со страницей.

In [98]:
top_devices = (
    df_filtered.groupby('device')['session_duration_min']
    .mean()
    .sort_values(ascending=False)
    .head(2)
    .index.tolist()
)

print("Топ-2 устройства по средней длительности сессии:", top_devices)

Топ-2 устройства по средней длительности сессии: ['Android', 'iPhone']


C:\Users\miros\AppData\Local\Temp\ipykernel_12740\2564763962.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_filtered.groupby('device')['session_duration_min']


Находим какие устройства являются топ-2 по длительности сессии - это Android и Iphone.

In [101]:
result = (
    df_filtered
    .groupby(['channel', 'device'])['session_duration_min']
    .median()
    .round(2)
    .reset_index()
    .rename(columns={'session_duration_min': 'Медианная_длительность'})
)

result

C:\Users\miros\AppData\Local\Temp\ipykernel_12740\824012485.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  .groupby(['channel', 'device'])['session_duration_min']


channel   device  Медианная_длительность
0       FaceBoom  Android                    36.0
1       FaceBoom      Mac                    26.0
2       FaceBoom       PC                    23.0
3       FaceBoom   iPhone                    25.0
4   MediaTornado  Android                    24.0
5   MediaTornado      Mac                    37.5
6   MediaTornado       PC                     NaN
7   MediaTornado   iPhone                    35.5
8         TipTop  Android                    27.0
9         TipTop      Mac                    26.0
10        TipTop       PC                    30.5
11        TipTop   iPhone                    32.0
12       organic  Android                    30.0
13       organic      Mac                    29.5
14       organic       PC                    32.0
15       organic   iPhone                    30.0

Создали сводную таблицу - канал, устройство и медианная длительность для отфильтрованного набора.

---
**Вывод:** Данная сводная таблица позволяет проанализировать какие рекламные каналы приносят активных пользователей и какие из них имеют наивысшую эффективность. Самый качественный трафик приходит из канала organic.
***